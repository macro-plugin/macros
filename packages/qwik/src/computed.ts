import { ArrowFunctionExpression, Expression, ExpressionStatement, Identifier, VariableDeclaration } from "@swc/core";
import { BaseNode, createLabeledMacro, markedNode, unMarkNode, walk } from "@macro-plugin/core";

export const computed = createLabeledMacro('computed', function (stmt) {
  if (stmt.type != 'BlockStatement') return;

  const computeds: Record<string, { value: BaseNode, computed?: BaseNode | Expression }> = {};
  const signals: string[] = [];

  let name;
  for (const i of stmt.stmts) {
    if (i.type == 'VariableDeclaration' && i.kind == 'var') {
      for (const d of i.declarations) {
        if (d.id.type == 'Identifier') {
          name = d.id.value;

          if (d.init) {
            computeds[name] = { value: JSON.parse(JSON.stringify(d.init)), computed: d.init };
            const isSignal = (name: string) => this.track(name)?.marker == 'qwikSignal';

            walk(d.init, {
              enter(node) {
                if (node.type == "Identifier") {
                  const name = (node as Identifier).value;
                  if (isSignal(name)) {
                    unMarkNode(node);
                    (node as Identifier).value = '__' + name;
                    if (!signals.includes(name)) signals.push(name);
                  }
                }
              },
            })
          }

        } else {
          throw new Error("Expect an Identifier")
        }
      }
    } else {
      throw new Error('Expect a `var` kind VariableDeclaration node in signal block')
    }
  }

  if (Object.keys(computeds).length > 0) {
    this.import([{ name: 'useSignal' }, { name: 'useTask$' }], '@builder.io/qwik');
    return [...Object.entries(computeds).map(([k, v]) => ({
      type: "VariableDeclaration",
      kind: "const",
      declarations: [
        {
          type: 'VariableDeclarator',
          id: markedNode('qwikSignal', {
            type: 'Identifier',
            value: k,
            span: {
              start: 0,
              end: 0,
              ctxt: 0
            },
            optional: false
          }),
          init: {
            type: 'CallExpression',
            callee: {
              type: 'Identifier',
              value: 'useSignal',
              span: {
                start: 0,
                end: 0,
                ctxt: 0
              },
              optional: false
            },
            arguments: [
              {
                expression: v.value
              }
            ],
            span: {
              start: 0,
              end: 0,
              ctxt: 0
            }
          },
          span: {
            start: 0,
            end: 0,
            ctxt: 0
          }
        }
      ],
      declare: false,
      span: {
        start: 0,
        end: 0,
        ctxt: 0
      }
    } as VariableDeclaration)), {
      type: "ExpressionStatement",
      span: {
        start: 0,
        end: 0,
        ctxt: 0
      },
      expression: {
        type: "CallExpression",
        callee: {
          type: "Identifier",
          value: "useTask$",
          optional: false,
          span: {
            start: 0,
            end: 0,
            ctxt: 0
          }
        },
        span: {
          start: 0,
          end: 0,
          ctxt: 0
        },
        arguments: [
          {
            expression: {
              type: "ArrowFunctionExpression",
              async: false,
              generator: false,
              params: [
                {
                  type: "ObjectPattern",
                  optional: false,
                  span: {
                    start: 0,
                    end: 0,
                    ctxt: 0
                  },
                  properties: [
                    {
                      type: "AssignmentPatternProperty",
                      span: {
                        start: 0,
                        end: 0,
                        ctxt: 0
                      },
                      key: {
                        type: "Identifier",
                        value: "track",
                        optional: false,
                        span: {
                          start: 0,
                          end: 0,
                          ctxt: 1
                        },
                      },
                    }
                  ]
                }
              ],
              body: {
                type: "BlockStatement",
                span: {
                  start: 0,
                  end: 0,
                  ctxt: 0
                },
                stmts: [
                  ...signals.map(name => ({
                    type: "VariableDeclaration",
                    kind: "const",
                    declarations: [
                      {
                        type: 'VariableDeclarator',
                        definite: false,
                        span: {
                          start: 0,
                          end: 0,
                          ctxt: 0
                        },
                        id: {
                          type: 'Identifier',
                          value: '__' + name,
                          span: {
                            start: 0,
                            end: 0,
                            ctxt: 0
                          },
                          optional: false
                        },
                        init: {
                          type: 'CallExpression',
                          callee: {
                            type: 'Identifier',
                            value: 'track',
                            span: {
                              start: 0,
                              end: 0,
                              ctxt: 0
                            },
                            optional: false
                          },
                          span: {
                            start: 0,
                            end: 0,
                            ctxt: 0
                          },
                          arguments: [
                            {
                              expression: {
                                type: "ArrowFunctionExpression",
                                async: false,
                                generator: false,
                                params: [],
                                body: {
                                  type: "Identifier",
                                  value: name,
                                  optional: false,
                                  span: {
                                    start: 0,
                                    end: 0,
                                    ctxt: 0
                                  }
                                },
                                span: {
                                  start: 0,
                                  end: 0,
                                  ctxt: 0
                                }
                              }
                            }
                          ]
                        }
                      }
                    ],
                    declare: false,
                    span: {
                      start: 0,
                      end: 0,
                      ctxt: 0
                    }
                  } as VariableDeclaration)),
                  ...Object.entries(computeds).map(([k, v]) => ({
                    type: "ExpressionStatement",
                    expression: {
                      type: "AssignmentExpression",
                      operator: "=",
                      left: markedNode('qwikSignal', {
                        type: 'Identifier',
                        value: k,
                        optional: false,
                        span: {
                          start: 0,
                          end: 0,
                          ctxt: 0
                        }
                      }),
                      right: v.computed,
                      span: {
                        start: 0,
                        end: 0,
                        ctxt: 0
                      }
                    },
                    span: {
                      start: 0,
                      end: 0,
                      ctxt: 0
                    }
                  } as ExpressionStatement))
                ]
              },
              span: {
                start: 0,
                end: 0,
                ctxt: 0
              }
            } as ArrowFunctionExpression
          }
        ]
      }
    }]
  }
})
