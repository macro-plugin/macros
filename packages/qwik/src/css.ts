import { CallExpression, ExpressionStatement } from "@swc/core";

import { createMacro } from "@macro-plugin/core";

export const css = createMacro({
  LabeledStatement(ast) {
    if (ast.label.value != 'css') return;
    if (ast.body.type == "ExpressionStatement") {
      const specifier = this.get('QwikScoped', false) ? 'useStyleScoped$' : 'useStyle$';

      this.import([{ name: specifier }], '@builder.io/qwik');

      ast.body.expression = {
        "type": "CallExpression",
        "span": {
          "start": 109,
          "end": 119,
          "ctxt": 0
        },
        "callee": {
          "type": "Identifier",
          "span": {
            "start": 109,
            "end": 112,
            "ctxt": 1
          },
          "value": specifier,
          "optional": false
        },
        "arguments": [
          {
            "expression": ast.body.expression
          }
        ],
      } as CallExpression

      return ast.body
    } else {
      throw new Error('this macro only accept an Expression')
    }
  }
})

export const link = createMacro({
  LabeledStatement(ast) {
    if (ast.label.value != 'link') return;
    if (ast.body.type != 'ExpressionStatement') return;
    let name: string;
    let linkCount = this.get('QwikLinkCount', 0);
    const scoped = this.get('QwikScoped', false);
    const specifier = scoped ? "useStyleScoped$" : "useStyles$";
    const links: string[] = [];
    if (ast.body.expression.type == 'StringLiteral') {
      name = '__link' + linkCount;
      links.push(name);
      this.import([{ name, kind: 'default' }], ast.body.expression.value);
      linkCount += 1;
    } else if (ast.body.expression.type == 'ArrayExpression') {
      for (const i of ast.body.expression.elements) {
        if (i?.expression.type == 'StringLiteral') {
          name = '__link' + linkCount
          links.push(name);
          this.import([{ name, kind: 'default' }], i.expression.value);
          linkCount += 1;
        } else {
          throw new Error("Expect a StringLiteral")
        }
      }
    } else {
      throw new Error("Only support StringLiteral or ArrayExpression")
    }
    this.set('QwikLinkCount', linkCount);
    this.import([{ name: specifier }], "@builder.io/qwik")
    return links.map(i => ({
      "type": "ExpressionStatement",
      "span": {
        "start": 109,
        "end": 119,
        "ctxt": 0
      },
      "expression": {
        "type": "CallExpression",
        "span": {
          "start": 109,
          "end": 119,
          "ctxt": 0
        },
        "callee": {
          "type": "Identifier",
          "span": {
            "start": 109,
            "end": 112,
            "ctxt": 1
          },
          "value": specifier,
          "optional": false
        },
        "arguments": [
          {
            "expression": {
              "type": "Identifier",
              "span": {
                "start": 113,
                "end": 118,
                "ctxt": 1
              },
              "value": i,
              "optional": false
            }
          }
        ],
      }
    } as ExpressionStatement))
  }
})

export const scoped = createMacro({
  LabeledStatement: {
    enter(ast) {
      if (ast.label.value != 'scoped') return;
      if (ast.body.type == 'BlockStatement') {
        this.set('QwikScoped', true);
      } else {
        throw new Error("Only accept BlockStatement in scoped macro.")
      }
    },
    leave(ast) {
      if (ast.label.value != 'scoped') return;
      if (ast.body.type == 'BlockStatement') {
        this.replace(ast.body.stmts);
        this.set('QwikScoped', false);
      }
    }
  }
})
