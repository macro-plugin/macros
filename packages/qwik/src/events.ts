import { ArrowFunctionExpression, ExpressionStatement } from "@swc/core"

import { createMacro } from "@macro-plugin/core"

export const events = createMacro({
  LabeledStatement: {
    enter (ast) {
      if (ast.label.value === "document" && ast.body.type === "BlockStatement") {
        this.set("QwikInDocument", true)
        return
      }

      if (ast.label.value === "window" && ast.body.type === "BlockStatement") {
        this.set("QwikInWindow", true)
        return
      }

      const inDoc = this.get("QwikInDocument", false)
      const inWin = this.get("QwikInWindow", false)

      if (!ast.label.value.startsWith("on") || !["BlockStatement", "ExpressionStatement"].includes(ast.body.type)) return

      let expression
      let name = inDoc ? "useOnDocument" : inWin ? "useOnWindow" : "useOn"
      this.import([{ name: "$" }], "@builder.io/qwik")

      if (ast.body.type === "BlockStatement") {
        const firstLabel = ast.body.stmts[0]
        if (firstLabel && firstLabel.type === "LabeledStatement") {
          if (firstLabel.label.value === "document") {
            ast.body.stmts = ast.body.stmts.slice(1)
            name = "useOnDocument"
          } else if (firstLabel.label.value === "window") {
            ast.body.stmts = ast.body.stmts.slice(1)
            name = "useOnWindow"
          }
        }

        expression = {
          type: "ArrowFunctionExpression",
          span: {
            start: 26,
            end: 87,
            ctxt: 0
          },
          params: [],
          body: ast.body,
          async: false,
          generator: false,
        } as ArrowFunctionExpression
      } else if (ast.body.type === "ExpressionStatement") {
        expression = ast.body.expression

        if (ast.body.expression.type === "FunctionExpression" || ast.body.expression.type === "ArrowFunctionExpression") {
          if (ast.body.expression.body?.type === "BlockStatement") {
            const firstLabel = ast.body.expression.body.stmts[0]
            if (firstLabel && firstLabel.type === "LabeledStatement") {
              if (firstLabel.label.value === "document") {
                ast.body.expression.body.stmts = ast.body.expression.body.stmts.slice(1)
                name = "useOnDocument"
              } else if (firstLabel.label.value === "window") {
                ast.body.expression.body.stmts = ast.body.expression.body.stmts.slice(1)
                name = "useOnWindow"
              }
            }
          }
        }
      }

      this.import([{ name }], "@builder.io/qwik")

      return {
        type: "ExpressionStatement",
        span: {
          start: 0,
          end: 93,
          ctxt: 0
        },
        expression: {
          type: "CallExpression",
          span: {
            start: 0,
            end: 92,
            ctxt: 0
          },
          callee: {
            type: "Identifier",
            span: {
              start: 0,
              end: 5,
              ctxt: 1
            },
            value: name,
            optional: false
          },
          arguments: [
            {
              expression: {
                type: "StringLiteral",
                span: {
                  start: 11,
                  end: 18,
                  ctxt: 0
                },
                value: ast.label.value.slice(2),
                raw: `'${ast.label.value.slice(2)}'`
              }
            },
            {
              expression: {
                type: "CallExpression",
                span: {
                  start: 24,
                  end: 88,
                  ctxt: 0
                },
                callee: {
                  type: "Identifier",
                  span: {
                    start: 24,
                    end: 25,
                    ctxt: 1
                  },
                  value: "$",
                  optional: false
                },
                arguments: [
                  {
                    expression
                  }
                ],
              }
            }
          ],
        }
      } as ExpressionStatement
    },
    leave (ast) {
      if (ast.label.value === "document" && ast.body.type === "BlockStatement") {
        this.replace(ast.body.stmts)
        this.set("QwikInDocument", false)
      }

      if (ast.label.value === "window" && ast.body.type === "BlockStatement") {
        this.replace(ast.body.stmts)
        this.set("QwikInWindow", false)
      }
    }
  }
})
