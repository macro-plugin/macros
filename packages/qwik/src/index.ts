export { qwik } from "./qwik"
export { task, vtask } from "./task"
export { resource } from "./resource"
export { store } from "./store"
export { signal } from "./signal"
export { events } from "./events"
export { computed } from "./computed"
export { link, css, scoped } from "./css"
