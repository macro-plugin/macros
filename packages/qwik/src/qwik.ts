import { ArrowFunctionExpression, CallExpression, FunctionDeclaration, FunctionExpression, LabeledStatement, VariableDeclaration } from "@swc/core"
import { WalkContext, createMacro, dummySpan } from "@macro-plugin/core"

function handleFunc (this: WalkContext, ast: FunctionDeclaration | FunctionExpression | ArrowFunctionExpression) {
  let label
  if (!ast.body || ast.body.type !== "BlockStatement") return
  for (let i = 0; i < ast.body.stmts.length; i++) {
    if (ast.body.stmts[i].type === "LabeledStatement") {
      label = ast.body.stmts[i] as LabeledStatement
      if (label.label.value === "qwik") {
        if (label.body.type === "ExpressionStatement" && label.body.expression.type === "BooleanLiteral") {
          ast.body.stmts = ast.body.stmts.filter((_, index) => index !== i)
          const isDecl = "identifier" in ast && ast.identifier != null

          if (label.body.expression.value) {
            this.import("component$", "@builder.io/qwik")

            const init = {
              type: "CallExpression",
              span: dummySpan,
              callee: {
                type: "Identifier",
                span: dummySpan,
                value: "component$",
                optional: false
              },
              arguments: [
                {
                  expression: isDecl
                    ? {
                      type: "ArrowFunctionExpression",
                      span: dummySpan,
                      params: ast.params.map(i => i.pat),
                      body: ast.body,
                      async: false,
                      generator: false,
                    }
                    : ast
                }
              ],
            } as CallExpression

            if (!isDecl) return init

            return {
              type: "VariableDeclaration",
              span: dummySpan,
              kind: "const",
              declare: false,
              declarations: [
                {
                  type: "VariableDeclarator",
                  span: dummySpan,
                  id: ast.identifier,
                  init,
                  definite: false
                }
              ]
            } as VariableDeclaration
          }
        } else {
          throw new Error("Expect an boolean in qwik macro.")
        }
      }
    }
  }
}

export const qwik = createMacro({
  FunctionDeclaration (ast) {
    return handleFunc.apply(this, [ast])
  },
  FunctionExpression (ast) {
    return handleFunc.apply(this, [ast])
  },
  ArrowFunctionExpression (ast) {
    return handleFunc.apply(this, [ast])
  }
})
