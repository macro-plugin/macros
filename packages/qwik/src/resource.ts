import { Identifier, Statement, VariableDeclaration } from "@swc/core";

import { createPlugin } from "@macro-plugin/core";

const varToReturn = (body: Statement[]) => {
  let ident: Identifier | undefined;
  body = body.map(i => {
    if (i.type == "VariableDeclaration" && i.kind == 'var') {
      if (i.declarations[0].id.type == 'Identifier') {
        ident = i.declarations[0].id;
      } else {
        throw new Error('Expect an Identifier')
      }
      return {
        type: 'ReturnStatement',
        argument: i.declarations[0].init,
        span: {
          start: 0,
          end: 0,
          ctxt: 0
        }
      }
    }
    return i
  })

  return [ident, body] as [Identifier, Statement[]]
}

export const resource = createPlugin(function (ast) {
  if (ast.type != 'LabeledStatement' || ast.label.value != 'resource') return;

  this.import([{ name: 'useResource$' }], '@builder.io/qwik');
  if (ast.body.type == "BlockStatement") {
    const [id, stmts] = varToReturn(ast.body.stmts);

    return {
      type: "VariableDeclaration",
      declarations: [
        {
          type: "VariableDeclarator",
          definite: false,
          span: {
            start: 0,
            end: 0,
            ctxt: 0
          },
          id,
          init: {
            type: 'CallExpression',
            callee: {
              type: 'Identifier',
              optional: false,
              value: 'useResource$',
              span: {
                start: 0,
                end: 0,
                ctxt: 0
              }
            },
            span: {
              start: 0,
              end: 0,
              ctxt: 0
            },
            arguments: [
              {
                expression: {
                  type: 'ArrowFunctionExpression',
                  generator: false,
                  async: false,
                  params: [],
                  span: {
                    start: 0,
                    end: 0,
                    ctxt: 0
                  },
                  body: {
                    type: 'BlockStatement',
                    stmts,
                    span: {
                      start: 0,
                      end: 0,
                      ctxt: 0
                    },
                  }
                }
              }
            ]
          }
        }
      ],
      span: {
        start: 0,
        end: 0,
        ctxt: 0
      },
      declare: false,
      kind: "const"
    } as VariableDeclaration
  } else if (ast.body.type == "ExpressionStatement" && ast.body.expression.type == 'ArrowFunctionExpression') {
    let id, stmts;

    if (ast.body.expression.body.type == 'BlockStatement') {
      [id, stmts] = varToReturn(ast.body.expression.body.stmts);
      ast.body.expression.body.stmts = stmts;
    } else {
      throw new Error('expect an arrow block inside resource.')
    }

    return {
      type: "VariableDeclaration",
      declarations: [
        {
          type: "VariableDeclarator",
          definite: false,
          span: {
            start: 0,
            end: 0,
            ctxt: 0
          },
          id,
          init: {
            type: 'CallExpression',
            callee: {
              type: 'Identifier',
              value: 'useResource$',
              span: {
                start: 0,
                end: 0,
                ctxt: 0
              },
              optional: false
            },
            arguments: [
              {
                expression: ast.body.expression
              }
            ],
            span: {
              start: 0,
              end: 0,
              ctxt: 0
            },
          }
        }
      ],
      kind: "const",
      span: {
        start: 0,
        end: 0,
        ctxt: 0
      },
      declare: false,
    } as VariableDeclaration
  } else {
    throw new Error('this macro only accept a ArrowFunction or BlockStatement')
  }
})
