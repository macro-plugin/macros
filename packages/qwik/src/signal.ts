import { MemberExpression, Node, VariableDeclaration } from "@swc/core";
import { createPlugin, markedNode } from "@macro-plugin/core";

const plugin = createPlugin({
  LabeledStatement(ast) {
    if (ast.body.type != 'BlockStatement' || ast.label.value != 'signal') return;

    const signals: Record<string, { value?: Node }> = {};

    let name;
    for (const i of ast.body.stmts) {
      if (i.type == 'VariableDeclaration' && i.kind == 'var') {
        for (const d of i.declarations) {
          if (d.id.type == 'Identifier') {
            name = d.id.value;
            signals[name] = { value: d.init };
          } else {
            throw new Error("Expect Identifier in signal")
          }
        }
      } else {
        throw new Error('Expect a `var` kind VariableDeclaration node in signal block')
      }
    }

    if (Object.keys(signals).length > 0) {
      this.import([{ name: 'useSignal' }], '@builder.io/qwik');
      return Object.entries(signals).map(([k, v]) => ({
        type: "VariableDeclaration",
        kind: "const",
        declare: false,
        declarations: [
          {
            type: 'VariableDeclarator',
            id: markedNode('qwikSignal', {
              type: 'Identifier',
              value: k,
              optional: false,
              span: {
                start: 0,
                end: 0,
                ctxt: 1
              }
            }),
            init: {
              type: 'CallExpression',
              callee: {
                type: 'Identifier',
                value: 'useSignal',
                optional: false,
                span: {
                  start: 0,
                  end: 0,
                  ctxt: 0
                }
              },
              arguments: [
                {
                  expression: v.value
                }
              ],
              span: {
                start: 0,
                end: 0,
                ctxt: 0
              }
            },
            definite: false,
            span: {
              start: 0,
              end: 0,
              ctxt: 0
            }
          }
        ],
        span: {
          start: 0,
          end: 0,
          ctxt: 0
        }
      } as VariableDeclaration))
    }

  },
  Identifier(ast, parent) {
    if (parent?.type != 'VariableDeclarator' && this.track(ast.value)?.marker == 'qwikSignal') {
      this.replace({
        type: 'MemberExpression',
        object: ast,
        property: {
          type: "Identifier",
          value: "value",
          optional: false,
          span: {
            start: 0,
            end: 0,
            ctxt: 0
          }
        },
        span: {
          start: 0,
          end: 0,
          ctxt: 0
        }
      } as MemberExpression);
      this.skip();
    }
  }
})

export default plugin;
