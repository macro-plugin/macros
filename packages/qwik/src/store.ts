import { Node, VariableDeclaration } from "@swc/core";
import { createMacro, markedNode } from "@macro-plugin/core";

export const store = createMacro({
  LabeledStatement(ast) {
    if (ast.body.type != 'BlockStatement' || ast.label.value != 'store') return;

    const signals: Record<string, { value?: Node }> = {};

    let name;
    for (const i of ast.body.stmts) {
      if (i.type == 'VariableDeclaration' && i.kind == 'var') {
        for (const d of i.declarations) {
          if (d.id.type == 'Identifier') {
            name = d.id.value;
            signals[name] = { value: d.init };
          } else {
            throw new Error("Expect Identifier in store")
          }
        }
      } else {
        throw new Error('Expect a `var` kind VariableDeclaration node in store block')
      }
    }

    if (Object.keys(signals).length > 0) {
      this.import([{ name: 'useStore' }], '@builder.io/qwik');
      return Object.entries(signals).map(([k, v]) => ({
        type: "VariableDeclaration",
        kind: "const",
        declare: false,
        declarations: [
          {
            type: 'VariableDeclarator',
            id: markedNode('qwikStore', {
              type: 'Identifier',
              value: k,
              optional: false,
              span: {
                start: 0,
                end: 0,
                ctxt: 1
              }
            }),
            init: {
              type: 'CallExpression',
              callee: {
                type: 'Identifier',
                value: 'useStore',
                optional: false,
                span: {
                  start: 0,
                  end: 0,
                  ctxt: 0
                }
              },
              arguments: [
                {
                  expression: v.value
                }
              ],
              span: {
                start: 0,
                end: 0,
                ctxt: 0
              }
            },
            definite: false,
            span: {
              start: 0,
              end: 0,
              ctxt: 0
            }
          }
        ],
        span: {
          start: 0,
          end: 0,
          ctxt: 0
        }
      } as VariableDeclaration))
    }
  },
})
