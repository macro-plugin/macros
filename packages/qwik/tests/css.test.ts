import { css, link, qwik, scoped } from "../src"

import { transform } from "@macro-plugin/core"

test("qwik css link", () => {
  const code = `
    export const CmpStyles = () => {
      qwik: true
      link: './code-block.css?inline'

      return <span class="my-text">Some text</span>
    }
  `
  expect(transform(code, { macros: [qwik, link], jsc: { parser: { syntax: "typescript", tsx: true } } }).code).toMatchSnapshot()
})

test("qwik css links", () => {
  const code = `
    export const CmpStyles = () => {
      qwik: true
      link: [
        './code-block.css?inline',
        './font-style.css?inline',
      ]

      return <span class="my-text">Some text</span>
    }
  `
  expect(transform(code, { macros: [qwik, link], jsc: { parser: { syntax: "typescript", tsx: true } } }).code).toMatchSnapshot()
})

test("qwik css string", () => {
  const code = `
    export const CmpStyles = () => {
      qwik: true
      css: \`
        .my-text {
          color: red;
        }
      \`

      return <span class="my-text">Some text</span>
    }
  `
  expect(transform(code, { macros: [qwik, css], jsc: { parser: { syntax: "typescript", tsx: true } } }).code).toMatchSnapshot()
})

test("qwik scoped css", () => {
  const code = `
  export const CmpStyles = () => {
    qwik: true
    scoped: {
      link: './code-block.css?inline'

      css: \`
        .my-text {
          color: red;
        }
      \`
    }

    return <span class="my-text">Some text</span>
  }
`
  expect(transform(code, { macros: [qwik, scoped, css, link], jsc: { parser: { syntax: "typescript", tsx: true } } }).code).toMatchSnapshot()
})
