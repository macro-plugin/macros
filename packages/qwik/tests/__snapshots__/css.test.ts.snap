// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`qwik css link 1`] = `
"import { component$ } from "@builder.io/qwik";
import __link0 from "./code-block.css?inline";
import { useStyles$ } from "@builder.io/qwik";
export const CmpStyles = component$(()=>{
    useStyles$(__link0);
    return <span class="my-text">Some text</span>;
});
"
`;

exports[`qwik css links 1`] = `
"import { component$ } from "@builder.io/qwik";
import __link0 from "./code-block.css?inline";
import __link1 from "./font-style.css?inline";
import { useStyles$ } from "@builder.io/qwik";
export const CmpStyles = component$(()=>{
    useStyles$(__link0);
    useStyles$(__link1);
    return <span class="my-text">Some text</span>;
});
"
`;

exports[`qwik css string 1`] = `
"import { component$ } from "@builder.io/qwik";
import { useStyle$ } from "@builder.io/qwik";
export const CmpStyles = component$(()=>{
    useStyle$(\`
        .my-text {
          color: red;
        }
      \`);
    return <span class="my-text">Some text</span>;
});
"
`;

exports[`qwik scoped css 1`] = `
"import { component$ } from "@builder.io/qwik";
import __link0 from "./code-block.css?inline";
import { useStyleScoped$ } from "@builder.io/qwik";
export const CmpStyles = component$(()=>{
    useStyleScoped$(__link0);
    useStyleScoped$(\`
        .my-text {
          color: red;
        }
      \`);
    return <span class="my-text">Some text</span>;
});
"
`;
